File: 4/lab4.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;

public struct Token
{
    public string data;
    public int recipient;
    public Token(string p1, int p2)
    {
        data = p1;
        recipient = p2;
    }
}
namespace ConsoleApplication3
{
    class Program
    {
        static void Main(string[] args)
        {
            Token token = new Token("data", 10);
            Thread thread = new Thread(() => ThreadFunction(1, token));
            thread.Start();
        }
        static void ThreadFunction(int i, Token t)
        {
            if (t.recipient == i)
            {
                Console.WriteLine("data:" + t.data + " recipient:" + i);
            }
            else
            {
                Thread thread = new Thread(() => ThreadFunction(i + 1, t));
                thread.Start();
            }
        }
    }
}

File: 4/lab4.go
package main
import (
	"fmt"
	"time"
)

type Token struct {
	data string
	recipient int
}

func main() {
	token := Token{data: "token", recipient: 10}
	go thread(1, token)
	time.Sleep(100 * time.Millisecond)
}

func thread(i int, t Token) {
	if (t.recipient == i) {
		fmt.Println("data:", t.data," recipient:",i)
	} else {
		go thread(i+1, t)
	}
}